### PR TITLE
Add context free URLs section

### LINKED ISSUE #28: URLs should be context-free
Based on discussion in Singapore, the group decided to recommend (not require) that URLs use names that can be resolved from the public DNS (no split horizon, etc...).  The content served from those URLs should also not be dependent on the network path, but instead the URL would include all information necessary to produce the resource.

### PR DESCRIPTION
**Pull request by liucougar**
Closes #28

### CODE REVIEW

File: draft-ietf-capport-architecture.xml [hunk 1]
```diff
@@ -695,7 +695,27 @@ o . . . . . . . . . . . . . . . . . . .| |. . . . . . . . . . . o
...
                 to which the IP belongs.
             </t>
-           </section>
+         </section>
+         <section anchor="context_free_url" title="Context-free URL">
+           <t>
+             The URLs provided in the API SHOULD contain all the information necessary to render
+             the resources requested: they should be context free so it does not depend on ambient
+             information, such as remote address on the connection. This is to ensure that if
+             User Equipment access a URL from another network path, the correct resource can be
```
Comment (martinthomson) [inline draft-ietf-capport-architecture.xml]: Maybe:

> This is to ensure that User Equipment can access URLs from a network other than the one that contains the captive portal.

File: draft-ietf-capport-architecture.xml [hunk 2]
```diff
@@ -695,7 +695,27 @@ o . . . . . . . . . . . . . . . . . . .| |. . . . . . . . . . . o
...
+             retrieved. One possible approach is to include the identification of the User
+             Equipment in the URLs.
+           </t>
+           <t>
+             The URLs provided in the API are recommended to be resolved using public DNS, so they
+             can be rendered in case User Equipment is making request from a different network.
+           </t>
+           <t>
+             Having User Equipment identification in the URL may expose the service to spoofing or
```
Comment (martinthomson) [inline draft-ietf-capport-architecture.xml]: Maybe you can also say "Though a URL might still correctly resolve when the UE makes the request from a different network, it is possible that some functions could be limited to when the UE makes requests using the captive network."

Comment (ekline) [reply draft-ietf-capport-architecture.xml]: Do we also need to have some text (possibly in another section) about the wisdom of having UE identifying information in the URL, especially if these URLs might be accessed from off-captive-net?  (one option is to include an authenticated hash of relevant bits or a just unique session that only the captive portal infrastructure can convert into UE-identifying information.)

File: draft-ietf-capport-architecture.xml (hunk 1 above)
Comment (liucougar) [reply draft-ietf-capport-architecture.xml]: changed

File: draft-ietf-capport-architecture.xml (hunk 2 above)
Comment (liucougar) [reply draft-ietf-capport-architecture.xml]: added this, thanks

File: draft-ietf-capport-architecture.xml [hunk 3]
```diff
@@ -695,7 +695,29 @@ o . . . . . . . . . . . . . . . . . . .| |. . . . . . . . . . . o
...
+           <t>
+             The URLs provided in the API SHOULD contain all the information necessary to render
+             the resources requested: they should be context free so it does not depend on ambient
+             information, such as remote address on the connection. This is to ensure that User
+             Equipment can access URLs from a network other than the one that contains the captive
+             portal. One possible approach is to include the identification of the User Equipment
+             in the URLs.
+           </t>
+           <t>
+             The URLs provided in the API are recommended to be resolved using public DNS, so they
```
Comment (klarose) [inline draft-ietf-capport-architecture.xml]: Should this be normative? (I.e. "RECOMMENDED")?

File: draft-ietf-capport-architecture.xml [hunk 4]
```diff
@@ -695,7 +695,29 @@ o . . . . . . . . . . . . . . . . . . .| |. . . . . . . . . . . o
                 link. In such cases, identification could be performed by subnet, such as the /64
                 to which the IP belongs.
             </t>
-           </section>
+         </section>
+         <section anchor="context_free_url" title="Context-free URL">
+           <t>
+             The URLs provided in the API SHOULD contain all the information necessary to render
+             the resources requested: they should be context free so it does not depend on ambient
```
Comment (klarose) [inline draft-ietf-capport-architecture.xml]: As phrased, this is a bit ambiguous. The resource should be context free, not the URL -- i.e. all context should be in the URL, and the method for generating the resource should not depend on anything but the context within that URL. In my first read of this section, I took this to mean that the UR: should be context free.

Perhaps rephrase to:

"...the resources requested: the resources should be context free so they do not ..."

File: draft-ietf-capport-architecture.xml [hunk 5]
```diff
@@ -695,7 +695,29 @@ o . . . . . . . . . . . . . . . . . . .| |. . . . . . . . . . . o
...
+             information, such as remote address on the connection. This is to ensure that User
+             Equipment can access URLs from a network other than the one that contains the captive
```
Comment (klarose) [inline draft-ietf-capport-architecture.xml]: Given my understanding (as outlined above), I think the issue isn't that the UE can't access URLs in a network other than the one in which they were initially captive, but rather that the resource at that URL may not render correctly in the new network. Perhaps some slight rephrasing could capture this?

**Comment by klarose** [comment]
Thanks for doing this! I've left a few minor comments.

File: draft-ietf-capport-architecture.xml (hunk 5 above)
Comment (ekline) [reply draft-ietf-capport-architecture.xml]: I could be mistaken, but my understanding of "access from another network" was that it could include something like a cell phone that tried to access the URL via the cell network and not via the captive wifi network.

Comment (klarose) [reply draft-ietf-capport-architecture.xml]: Yeah, that makes sense -- that's an entirely plausible scenario. I thought that the problem was that if the device serving the content at the URL used information from its network stack to render the content, then it would not work in this case.

E.g. imagine that it determined the identity of the UE based on the source ip address of the tcp connection over which the http request was carried. That would almost certainly fail if the UE switched from a captive wifi network to a cell-phone network. However, if the information the UE used to identify itself to the device in the first place was carried in the URL, this would not be a problem.

Comment (klarose) [reply draft-ietf-capport-architecture.xml]: My reading of the current text is that it's saying that we want to put to context into the URL so that the URL is still accessible. But, if we follow the advice of using a public DNS name for the host, it should always be accessible. At issue is the content it serves being meaningful.

File: draft-ietf-capport-architecture.xml (hunk 4 above)
Comment (liucougar) [reply draft-ietf-capport-architecture.xml]: how about:
...the resources requested: the resources should not depend on ambient...

File: draft-ietf-capport-architecture.xml (hunk 5 above)
Comment (liucougar) [reply draft-ietf-capport-architecture.xml]: I guess we have different interpretations of context

personally, I would not call anything in the URL as the context of the URL

for example, https://www.google.com/?q=ietf, the search query is part of the URL (and I would not call that context of the url)

I think the current text is stating that the resource behind the URL should be accessible even if a non-captive network is used to access the URL. 

Shall we clarify what context means here? or what do you suggest rephrase please?

Comment (klarose) [reply draft-ietf-capport-architecture.xml]: I think we agree on what the context is. Just for good measure, I'll give an example. :)

For example, if the API returned https://example.com/captive-portal for its `user-portal-url`, and example.com used the source-ip address of the connection to present quota information to the user while rendering that page, we'd have a problem if the user briefly switched to an LTE network from the captive wifi network. 

However, if the API returned https://example.com/captive-portal?ue-identifier=X, and example.com used X to look up and present the quota information, then it wouldn't matter which network the UE was connected to.

My concern was with the phrasing:

> This is to ensure that User Equipment can access URLs from a network other"

The question of whether the URL is *accessible* is independent of whether the content served at that URL is context free. What matters here is whether it's *meaningful* and *correct*. So, maybe rephrase to:

"This is to ensure that the content served from these URLs is correct and meaningful to the User Equipment, even when accessed from a network other than the one that contains the captive portal".

Sorry about the confusion and ramblings. Being somewhat disconnected from the WG at the moment, I just want to make sure I'm not misunderstanding. :)

File: draft-ietf-capport-architecture.xml (hunk 4 above)
Comment (klarose) [reply draft-ietf-capport-architecture.xml]: Thanks. That sounds good.

File: draft-ietf-capport-architecture.xml (hunk 3 above)
Comment (ekline) [reply draft-ietf-capport-architecture.xml]: This doc is currently targeting Informational.  We could all-caps recommend here, but it might be good to replicate the message over in the standards track API doc as well.

**Comment by martinthomson** [comment]
Just for transparency.  Erik and I discussed this and made a few editorial changes.

**Comment by liucougar** [comment]
thanks, was just trying to work on the changes here
